python/src/robyn/visualization/pareto_visualizer.py

```diff
@@ -234,23 +234,218 @@ def generate_diagnostic_plot(self, ax: Optional[plt.Axes] = None) -> Optional[pl
+        ax.set_ylabel(None)
+        ax.set_title('Immediate vs. Carryover Response Percentage')
+        
+        # Grid customization
+        ax.grid(True, axis='x', alpha=0.2)
+        ax.grid(False, axis='y')
+        ax.set_axisbelow(True)
+        
+        # Use white background
+        ax.set_facecolor('white')
+        
+        if fig:
+            plt.tight_layout()
+            return fig
+        return None
 
-    def generate_adstock_rate(self) -> plt.Figure:
-        """Generate plot showing adstock rates over time by channel.
+    def generate_adstock_rate(self, ax: Optional[plt.Axes] = None) -> Optional[plt.Figure]:
+        """Generate adstock rate visualization based on adstock type.
+        
+        Args:
+            ax: Optional matplotlib axes to plot on. If None, creates new figure
             
         Returns:
-            plt.Figure: Line plot of adstock decay rates
-        """
-
-        """
-            NOTE: Missing intervalType mapping in data_mapper from input collect.
+            plt.Figure if ax is None, else None
         """
-        fig, ax = plt.subplots()
\ No newline at end of file
+        # Get the plot data
+        plot_data = next(iter(self.pareto_result.plot_data_collect.values()))
+        adstock_data = plot_data['plot3data']
+        
+        # Create figure if no axes provided
+        if ax is None:
+            fig, ax = plt.subplots(figsize=(10, 8))
+        else:
+            fig = None
+        
+        # Handle different adstock types
+        if self.adstock == AdstockType.GEOMETRIC:
+            # Get geometric adstock data
+            dt_geometric = adstock_data['dt_geometric'].copy()
+            
+            # Create bar chart
+            bars = ax.barh(y=range(len(dt_geometric)), 
+                        width=dt_geometric['thetas'],
+                        height=0.5,
+                        color='coral')
+            
+            # Add percentage labels
+            for i, (theta) in enumerate(dt_geometric['thetas']):
+                ax.text(theta + 0.01, i,
+                    f"{theta*100:.1f}%",
+                    va='center',
+                    fontweight='bold')
+            
+            # Customize axes
+            ax.set_yticks(range(len(dt_geometric)))
+            ax.set_yticklabels(dt_geometric['channels'])
+            
+            # Format x-axis as percentage
+            ax.xaxis.set_major_formatter(plt.FuncFormatter(lambda x, p: f'{x*100:.0f}%'))
+            ax.set_xlim(0, 1)
+            
+            # Set title and labels
+            interval_type = self.mmm_data.mmmdata_spec.interval_type if self.mmm_data else "day"
+            ax.set_title('Geometric Adstock: Fixed Rate Over Time')
+            ax.set_xlabel(f'Thetas [by {interval_type}]')
+            ax.set_ylabel(None)
+            
+        elif self.adstock in [AdstockType.WEIBULL_CDF, AdstockType.WEIBULL_PDF]:
+            # Get Weibull data
```

Comment:
When refactoring in another PR, lets break out this buisiness logic into 2 separate functions which would plot their respective plots